repl: Support for eager evaluation

antsmartian opened the pull request:
Address the issue https://github.com/nodejs/node/issues/20977

~~Well, I would say its "initial" checkins. May be work will be required for handling cursors, additional~~ ~~edge cases (may be 🤔 ) etc, but I guess this should be a good starting point.~~

Added code for handling cursors and also for line refresh, when necessary. 

Thanks for your time on this. 

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Trott commented:
@nodejs/repl

devsnek commented:
any reason this is limited to call expressions?

antsmartian commented:
@devsnek I need to extend to other expressions as well, but just wanted to post my progress so that I can get review from the community.

antsmartian commented:
@devsnek Added support for other expressions and as well for `Identifier` too.

devsnek commented on lib/repl.js:
```diff
@@ -844,10 +845,61 @@ REPLServer.prototype.createContext = function() {
+    }
+
+    try {
+      if (isValidExpression(line)) {
```
there's no need to check if its a valid expression, or even an expression at all.

antsmartian replied:
Any reasons for that? You mean we can just wrap them in a try catch block and that should do good for us?

antsmartian replied:
I just thought sending code for every key press for eval might be unnecessary.

Fishrock123 commented on lib/readline.js:
```diff
@@ -456,6 +457,23 @@ Interface.prototype._insertString = function(c) {
+};
+
+// Refresh the line if preview present
+Interface.prototype.clearPreview = function() {
```
These should either be hidden (think: `Symbol` access), or documented. I'm not sure these make sense in readline, so maybe they should be in repl or private?

antsmartian replied:
I guess making it hidden with `Symbol` access would do good.

antsmartian commented:
@Fishrock123 Yes, you are correct. Not all edits (with valid code) trigger eager eval. I knew that it needs to be worked out. Regarding color will change them. Thanks for taking time to review this!

danbev commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/17313/

antsmartian commented:
Just to give an update, I will find some time later this week, afterwhich I will be updating the PR with comments addressed. Sorry for the delay :)

antsmartian commented:
@devsnek @Fishrock123 Addressed your comments. Now eager eval will work on all lines. Also changed the color to grey.

Let me know if anything else is needed. Thanks!

antsmartian commented:
Friendly remainder : @devsnek @Fishrock123

thefourtheye commented on lib/readline.js:
```diff
@@ -456,6 +458,28 @@ Interface.prototype._insertString = function(c) {
     // a hack to get the line refreshed if it's needed
     this._moveCursor(0);
+    // emit current line for generating preview
+    this.emit('buffer', this.line);
```
This line is there in both the branches, at the end. Perhaps this can be moved out of the `if..else`.

thefourtheye commented on lib/repl.js:
```diff
@@ -844,10 +844,54 @@ REPLServer.prototype.createContext = function() {
+        }, (error, result) => {
+
+          if (error) {
+            // mostly possible side effect exception
```
Its better to expect the specific exceptions and ignore them. Is that possible?

thefourtheye replied:
Also if there is an error, will the `result` object always have `result` property?

antsmartian replied:
Yes `result` will always have `result` even in case of `error. Will refactor the variable to `previewResult`, so that its better readable.

antsmartian replied:
And about the exception, it can be syntax issues, side effects etc. Not sure if I can catch them specifically. @devsnek can give some light on this.

thefourtheye commented on test/parallel/test-repl-preview-result.js:
```diff
@@ -0,0 +1,39 @@
+
+  assert.ok(actual[0].includes(wrapWithColorCode('\'test\'')));
+  // side effect scenario, preview is not available
+  assert.ok(!actual[2].includes('// '));
```
`appendPreview` doesn't have a space character after `//`. Is this intentional?

antsmartian replied:
Nope, it was a miss. Good catch. Will fix them.

antsmartian commented:
@thefourtheye I have addressed your comments. 

@devsnek @jdalton @Fishrock123  Can you please have a look?

Fishrock123 commented on lib/readline.js:
```diff
@@ -457,6 +458,28 @@ Interface.prototype._insertString = function(c) {
+};
+
+Interface.prototype.appendPreview = function(result) {
+  this.previewResult = `\u001b[90m //${result}\u001b[39m`;
```
```suggestion
  this.previewResult = `\u001b[90m // ${result}\u001b[39m`;
```

Fishrock123 commented on lib/repl.js:
```diff
@@ -844,10 +844,49 @@ REPLServer.prototype.createContext = function() {
   });
 
   addBuiltinLibsToObject(context);
+  previewResults.call(this);
```
This shouldn't be necessary - can you instead add the function on the class but either prefix the function name with `_` or, better, put it behind a Symbol? Something like `kPreviewResults = new Symbol('Preview Results Fn')`, and then you can do `this[kPreviewResults]()`.

Fishrock123 commented on lib/repl.js:
```diff
@@ -844,10 +844,49 @@ REPLServer.prototype.createContext = function() {
   return context;
 };
 
+function previewResults() {
```
Do we really need this wrapping function? Also, it's confusing, since there is a variable with the same name.

Fishrock123 commented on lib/repl.js:
```diff
@@ -844,10 +844,49 @@ REPLServer.prototype.createContext = function() {
+        }
+
+        if (undefined !== previewResult.result.value) {
+          const value = util.inspect(previewResult.result.value);
```
Are you sure this is what you think it is?

It currently seems to be printing Objects as Arrays of values:

```
> process.versions
{ http_parser: '2.8.0',
  node: '12.0.0-pre',
  v8: '7.0.276.38-node.11',
  uv: '1.24.0',
  zlib: '1.2.11',
  ares: '1.15.0',
  modules: '67',
  nghttp2: '1.34.0',
  napi: '3',
  openssl: '1.1.0i',
  icu: '63.1',
  unicode: '11.0',
  cldr: '34.0',
  tz: '2018e' }
> process.versions //[ '2.8.0', '12.0.0-pre', '7.0.276.38-node.11', '1.24.0', '1.2.11' ]
```

Fishrock123 commented on lib/repl.js:
```diff
@@ -844,10 +844,49 @@ REPLServer.prototype.createContext = function() {
+
+        if (previewResult.result.preview) {
+          const previewResults = previewResult.result.preview.properties;
+          const previewData = previewResults.map((preview) => preview.value); // eslint-disable-line max-len
```
I think this should avoid the eslint message:
```suggestion
          const previewData = previewResults.map(
            (preview) => preview.value
          );
```

Fishrock123 commented on lib/repl.js:
```diff
@@ -844,10 +844,49 @@ REPLServer.prototype.createContext = function() {
+          Interface.prototype.appendPreview.call(this, value);
+        }
+      });
+    }, () => []);
```
Did you mean this?
```suggestion
    }, () => {});
```

Fishrock123 commented on test/parallel/test-repl-preview-result.js:
```diff
@@ -0,0 +1,39 @@
+r.on('exit', common.mustCall(() => {
+  const actual = output.split('\n');
+
+  const wrapWithColorCode = (str) => `\u001b[90m //${str}\u001b[39m`;
```
```suggestion
  const wrapWithColorCode = (str) => `\u001b[90m // ${str}\u001b[39m`;
```

antsmartian commented:
@Fishrock123 Thanks for your review, will update the thread once I'm done with your review comments.

antsmartian commented:
@nodejs/repl Have added test cases and made changes requested by @Fishrock123

antsmartian commented:
Gentle remainder @nodejs/repl @Fishrock123 🔉

BridgeAR commented on lib/readline.js:
```diff
@@ -500,6 +501,28 @@ Interface.prototype._insertString = function(c) {
+  this.emit('buffer', this.line);
+};
+
+Interface.prototype.appendPreview = function(result) {
```
Are these function meant to be added to the prototype? If so, they should be documented.

antsmartian replied:
I have moved them to `_ appendPreview`, for which I guess docs are not needed. Yes, still it can be accessible in the prototype, but since this method is needed in other place (repl), I don't know how should I expose it (given the fact `_tabComplete` etc are working in this model, so should be fine).

BridgeAR commented on lib/readline.js:
```diff
@@ -500,6 +501,28 @@ Interface.prototype._insertString = function(c) {
+};
+
+Interface.prototype.appendPreview = function(result) {
+  this.previewResult = `\u001b[90m // ${result}\u001b[39m`;
```
Colors might not be supported. Please check that first before using them.

antsmartian replied:
Ah, I wasn't aware of this color part. Will check them and update the PR.

antsmartian replied:
Have added logic below to handle the color part, PTAL..

antsmartian commented:
@BridgeAR will address your comments.
@nodejs/repl @Fishrock123 Friendly ping. Can we get some reviews on this? I want to get this to a closure.. Its open for sometime

antsmartian commented:
@nodejs/repl @Fishrock123 @BridgeAR  PTAL..

antsmartian commented:
@BridgeAR Yes, I have addressed the comments and also added check for color part in the prev commit, may be you can have a look now. Thanks!

BridgeAR commented on lib/readline.js:
```diff
@@ -481,6 +482,42 @@ Interface.prototype._insertString = function(c) {
+
+  if (columns) {
+    const colorDepth = this.output.getColorDepth();
+    const s = colorDepth > 256 ?
```
The maximum color depth is 24. You can also use `hasColors()` instead of `getColorDepth()` as it's more intuitive to use.

BridgeAR commented on lib/readline.js:
```diff
@@ -481,6 +482,42 @@ Interface.prototype._insertString = function(c) {
+    const colorDepth = this.output.getColorDepth();
+    const s = colorDepth > 256 ?
+      `${this._prompt}${this.line}${this.previewResult}` :
+      `${this._prompt}${this.line}${result}`;
```
This is not identical to the color part (it does not only strip the colors).

BridgeAR commented on lib/readline.js:
```diff
@@ -481,6 +482,42 @@ Interface.prototype._insertString = function(c) {
+    const s = colorDepth > 256 ?
+      `${this._prompt}${this.line}${this.previewResult}` :
+      `${this._prompt}${this.line}${result}`;
+    this.output.write(s.length < columns ?
```
The length check includes the colors but that should ideally not be the case. Colors are not visible characters and should as such not be calculated towards the total length.

BridgeAR commented on lib/readline.js:
```diff
@@ -481,6 +482,42 @@ Interface.prototype._insertString = function(c) {
+      `${this._prompt}${this.line}${result}`;
+    this.output.write(s.length < columns ?
+      s : `${s.slice(0, columns - 3)
+        .replace(/\r?\n|\r/g, '')}...\u001b[39m`);
```
This adds the color unconditionally. Can you also please elaborate what exactly this does? I am not sure I understand why this part is sliced and replaced.

antsmartian replied:
The slice and replace part is mainly for splitting the new lines. Since in output terminal has only limited columns, we need to split them and replace with empty space so that we can show them the content in single line.

BridgeAR replied:
We already use `util.inspect()` on the preview. We should use options that deactivate the colors for the preview and make sure the output is printed on a single line. That can be done with the `compact` options set to `true` in combination with `breakLength` set to `Infinity` and `colors` set to `false`.

I would then limit the preview only instead of limiting the input as well. Right now we unconditionally slice the input value and that doesn't seem right.

antsmartian replied:
@BridgeAR: (sorry for a very late reply)

What you are saying is actually a good idea. But I was just trying to play with the options provided by you but still that doesn't seems to be atleast fixing the problem what we have now. For example, if the user types `process`, we get a long string, so if we apply the options like below:

```
util.inspect(process, {breakLength: Infinity, compact: true})
```

still we are getting results like below:

<img width="891" alt="Screen Shot 2019-11-27 at 11 36 47 AM" src="https://user-images.githubusercontent.com/1241511/69698046-3aad4300-110a-11ea-81e6-a4031143e271.png">

This actually breaks the cursor position as the output is huge. I feel, the only way for us to slice the string to the length of the tty `column`. May be is there any other way we could achieve it? 🤔

BridgeAR commented on lib/readline.js:
```diff
@@ -481,6 +482,42 @@ Interface.prototype._insertString = function(c) {
+
+// Append eager eval result to the
+// Current line
+Interface.prototype._appendPreview = function(result) {
```
It would be possible to add internal functionality in a separate file that would only be loaded internally. That way we do not have to pollute the prototype further. An underscore in a property name would not hinder anyone from using or monkey patching this function.

antsmartian replied:
Agreed, I will move them to `internal/readline/util`.

BridgeAR commented on lib/repl.js:
```diff
@@ -226,6 +229,66 @@ function REPLServer(prompt,
+        if (!previewResult.exceptionDetails && previewResult.result.objectId) {
+          eagerSession.post('Runtime.callFunctionOn', {
+            functionDeclaration:
+              'function(arg) { return util.inspect(arg, {depth: Infinity}) }',
```
`depth: Infinity` does not sound like a good default to me. The preview should ideally be short and on the same line. This could cause the terminal to add lots of output.

BridgeAR commented on lib/readline.js:
```diff
@@ -481,6 +482,42 @@ Interface.prototype._insertString = function(c) {
+      s : `${s.slice(0, columns - 3)
+        .replace(/\r?\n|\r/g, '')}...\u001b[39m`);
+  } else {
+    this.output.write(this._prompt + this.line + this.previewResult);
```
This adds colors unconditionally.

BridgeAR commented on lib/readline.js:
```diff
@@ -481,6 +482,42 @@ Interface.prototype._insertString = function(c) {
+  const columns = this.output.columns;
+  const hasColors = this.output.hasColors();
+  const s = hasColors ?
+    `${this._prompt}${this.line}${this.previewResult}` : line;
```
I personally would keep the preview, even if no colors are supported. The comment should be indication enough.

If we do not want to have a preview in case the colors are deactivated, we should not start the `eagerSession` in the REPL. That way no code is run unnecessary.

And is it correct that `this._prompt` is necessary in one case but not in the other?

antsmartian replied:
Hmmm, I'm planning to send the output to this function and then use `util.inspect` within the function to customize our output. Do you think that would be a better approach? Because anyways I need to call `util.inspect` in my REPL session.

lundibundi commented:
ping @antsmartian, will you be working on this further?

antsmartian commented:
@lundibundi Yes. This is open for long time ( I know :( ), I wanted it to get closed. Will look at it this week. Will update the PR soon.

lundibundi commented on lib/repl.js:
```diff
@@ -265,6 +269,66 @@ function REPLServer(prompt,
+            executionContextId: eagerEvalContextId,
+            returnByValue: true,
+          }, (err, result) => {
+            if (!err) {
```
Nit: perhaps also `debug`-log an error IIUC?

lundibundi commented on lib/repl.js:
```diff
@@ -265,6 +269,66 @@ function REPLServer(prompt,
+
+        // If no exception and we have objectId
+        // Run the expression via callFunctionOn
+        // And return it from util inspect.
```
Nit:
```javascript
// If there was no exception and we've got an objectId
// stringify it with `util.inspect` via Runtime.callFunctionOn.
```

antsmartian commented:
@lundibundi @nodejs/repl PTAL.. 🔉

lundibundi commented on lib/internal/repl/utils.js:
```diff
@@ -87,7 +89,98 @@ function isRecoverableError(e, code) {
+    return readline;
+  };
+
+  const { cursorTo, clearScreenDown } = lazyReadline();
```
:thinking: can we just `const { cursorTo, clearScreenDown } = require('readline')` at the top now?

lundibundi commented on lib/internal/repl/utils.js:
```diff
@@ -87,7 +89,98 @@ function isRecoverableError(e, code) {
+    repl.output.write(s);
+  }
+
+  // Move back the cursor to the original position
```
Nit:
```suggestion
  // Move back the cursor to the original position.
```

lundibundi commented on lib/internal/repl/utils.js:
```diff
@@ -87,7 +89,98 @@ function isRecoverableError(e, code) {
 
+// Appends the preview of the result
+// to the tty.
+function appendPreview(repl, result, cursorTo, clearScreenDown) {
```
```suggestion
function appendPreview(repl, preview, cursorTo, clearScreenDown) {
```
maybe? To make it a little clearer.

lundibundi commented on lib/internal/repl/utils.js:
```diff
@@ -87,7 +89,98 @@ function isRecoverableError(e, code) {
 }
 
+// Appends the preview of the result
+// to the tty.
```
Nit: I think this will fit on one line.

lundibundi commented on lib/readline.js:
```diff
@@ -490,6 +491,8 @@ Interface.prototype._insertString = function(c) {
     // A hack to get the line refreshed if it's needed
     this._moveCursor(0);
   }
+  // Emit current line for generating preview
```
Nit:
```suggestion
  // Emit current line for generating preview.
```

lundibundi commented on lib/internal/repl/utils.js:
```diff
@@ -87,7 +89,98 @@ function isRecoverableError(e, code) {
+      return;
+    }
+
+    if (undefined !== previewResult.result.value) {
```
Nit:
```suggestion
    if (previewResult.result.value !== undefined) {
```

lundibundi commented on lib/repl.js:
```diff
@@ -223,6 +225,20 @@ function REPLServer(prompt,
+  const eagerSession = new inspector.Session();
+  eagerSession.connect();
+  eagerSession.once('Runtime.executionContextCreated',
+                    ({ params: { context } }) => {
```
Perhaps to avoid big indentation:
```js
eagerSession.once(
  'Runtime.executionContextCreated',
  ({ params: { context } }) => {
  
  })
```

lundibundi commented on lib/internal/repl/utils.js:
```diff
@@ -87,7 +89,98 @@ function isRecoverableError(e, code) {
+  if (columns !== undefined) {
+    repl.output.write(line.length < columns ?
+      s : `${s.slice(0, columns - 3)
+        .replace(/\r?\n|\r/g, '')}...\u001b[39m`);
```
Nit:
```js
    repl.output.write(line.length < columns ?
      s : s.slice(0, columns - 3).replace(/\r?\n|\r/g, '') + '...\u001b[39m');
```

BridgeAR commented:
This got superseded by #30811 
@antsmartian thank you very much for the awesome work!
